packages/safe-apps-react-sdk/src/index.tsx

```diff
@@ -1,34 +1,53 @@
-import React, { ReactNode } from 'react';
-import connectSafe, { Safe } from './safe';
+import { createContext, useState, useEffect, useContext, useMemo, ReactElement } from 'react';
+import SafeAppsSDK, { Opts as SDKOpts, SafeInfo } from '@gnosis.pm/safe-apps-sdk';
 
-const SafeContext = React.createContext<Safe | undefined>(undefined);
+type SafeReactSDKContext = {
+  sdk: SafeAppsSDK;
+  connected: boolean;
+  safe: SafeInfo;
+};
+
+const SafeContext = createContext<SafeReactSDKContext | undefined>(undefined);
 
 interface Props {
-  loading?: ReactNode;
+  loader?: ReactElement;
+  opts?: SDKOpts;
 }
 
-export const SafeProvider: React.FC<Props> = ({ loading, children }) => {
-  const [safe] = React.useState(connectSafe());
-  const [connected, setConnected] = React.useState(false);
-  React.useEffect(() => {
-    safe.activate(() => {
-      setConnected(safe.isConnected());
-    });
-
-    return () => safe.deactivate();
-  }, [safe]);
-
-  return (
-    <div className="App">
-      {connected ? <SafeContext.Provider value={safe}>{children}</SafeContext.Provider> : loading}
-    </div>
-  );
+export const SafeProvider: React.FC<Props> = ({ loader = null, opts, children }) => {
+  const [sdk] = useState(new SafeAppsSDK(opts));
+  const [connected, setConnected] = useState(false);
+  const [safe, setSafe] = useState<SafeInfo>({ safeAddress: '', network: 'rinkeby' });
+  const contextValue = useMemo(() => ({ sdk, connected, safe }), [sdk, connected, safe]);
+
+  useEffect(() => {
+    const fetchSafeInfo = async () => {
+      try {
+        const safeInfo = await sdk.getSafeInfo();
+        setConnected(true);
+        setSafe(safeInfo);
```

Comment:
Should we first call `setSafe` and then `setConnected` ? that that the hook always returns the correct Safe info when it is connected? (Not sure if it actually makes a difference, as JS is not truly concurrent)

Reply:
yeah looks like it makes a difference: https://stackoverflow.com/questions/53048495/does-react-batch-state-update-functions-when-using-hooks and https://www.reddit.com/r/reactjs/comments/dobo9p/does_react_still_batch_state_updates_with/